crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -358,29 +780,152 @@ impl QuiltApi<QuiltVersionV1> for QuiltV1 {
+    fn range_read_from_columns(
+        &self,
+        start_col: usize,
+        end_col: usize,
+        bytes_to_skip: usize,
+        bytes_to_return: usize,
+    ) -> Result<Vec<u8>, QuiltError> {
+        if self.symbol_size == 0 || self.row_size == 0 || self.data.is_empty() {
+            return Err(QuiltError::Other("empty quilt data".to_string()));
+        }
+
+        let total_cols_in_quilt = self.row_size / self.symbol_size;
+        if end_col > total_cols_in_quilt {
+            return Err(QuiltError::IndexOutOfBounds(end_col, total_cols_in_quilt));
+        }
 
-                blob[written..written + chunk_size].copy_from_slice(&self.data[start_idx..end_idx]);
-                written += chunk_size;
+        // Initialize state
+        let n_rows = self.data.len() / self.row_size;
+        let symbols_to_skip = bytes_to_skip / self.symbol_size;
+        let mut current_col = start_col + symbols_to_skip / n_rows;
+        let mut current_row = symbols_to_skip % n_rows;
+        let mut bytes_to_skip = bytes_to_skip - (symbols_to_skip * self.symbol_size);
+        let mut bytes_to_return = bytes_to_return;
+        let mut result = Vec::with_capacity(bytes_to_return);
+
+        // Helper function to calculate data slice for current position.
+        let get_slice = |col: usize, row: usize, skip: usize, limit: usize| {
+            let base_index = row * self.row_size + col * self.symbol_size;
+            let start_index = base_index + skip;
+            let end_index = (base_index + self.symbol_size)
+                .min(start_index + limit)
+                .min(self.data.len());
+            &self.data[start_index..end_index]
+        };
+
+        // Helper function to advance to the next position
+        let advance_position = |col: &mut usize, row: &mut usize| {
+            *row = (*row + 1) % n_rows;
+            if *row == 0 {
+                *col += 1;
             }
+        };
+
+        // Process cells until we've collected enough bytes or run out of columns.
+        while current_col < end_col && bytes_to_return > 0 {
+            let slice = get_slice(current_col, current_row, bytes_to_skip, bytes_to_return);
+            result.extend_from_slice(slice);
+            bytes_to_return -= slice.len();
+
+            advance_position(&mut current_col, &mut current_row);
+            bytes_to_skip = 0;
         }
-        Ok(blob)
+
+        Ok(result)
+    }
+}
+
+impl QuiltV1 {
+    /// Returns an iterator over the symbols for a given column range.
+    fn iter_symbols(
+        data: &[u8],
+        row_size: usize,
+        symbol_size: usize,
+        start_col: usize,
+        end_col: usize,
+        begin_position: usize,
+        limit: usize,
+    ) -> Result<QuiltSymbolIter<'_>, QuiltError> {
+        QuiltSymbolIter::new(
+            data,
+            row_size,
+            symbol_size,
+            start_col,
+            end_col,
+            begin_position,
+            limit,
+        )
+    }
```

Comment:
Question: Related to another question above, do we need an iterator here? It seems to me that something similar as `range_read_from_columns` would work here as well.